plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java

```diff
@@ -0,0 +1,218 @@
+
+  public int getProcessedPackages() {
+    return processedPackages;
+  }
+
+  public int getProcessedPackageVersions() {
+    return processedPackageVersions;
+  }
+
+  @Override
+  public void processItem(final WalkerContext context, final StorageItem item)
+      throws Exception
+  {
+    // nop
+  }
+
+  public void onCollectionExit(final WalkerContext context, final StorageCollectionItem coll)
+      throws Exception
+  {
+    final TreeMap<Version, Map<String, Object>> packages = Maps.newTreeMap();
+
+    // collect all the possible tarballs
+    final Collection<StorageItem> files = coll.list();
+    for (StorageItem file : files) {
+      if (!(file instanceof StorageFileItem)) {
+        // not a file, just skip it
+        continue;
+      }
+      if (!file.getName().endsWith(".tgz")) {
+        // not a tarball, just skip it
+        continue;
+      }
+      try {
+        final Map<String, Object> packageJson = extractPackageJson((StorageFileItem) file);
+        if (packageJson == null || !packageJson.containsKey("name") || !packageJson.containsKey("version")) {
+          // not found in tarball, not containing needed elements, just skip it but this time log it
+          log.info("Not found or malformed package.json from {}", file.getRepositoryItemUid());
+          continue;
+        }
+        final Version version = Version.parse(String.valueOf(packageJson.get("version")));
+        packages.put(version, packageJson);
+      }
+      catch (Exception e) {
+        log.info("Failed to extract or malformed package.json from {}", file.getRepositoryItemUid());
+      }
+    }
+
+    // ie. dir empty or all extract failed?
+    if (packages.isEmpty()) {
+      return;
+    }
+
+    // rebuild package root as possible and push it into database
+    final Map<String, Object> rootJson = Maps.newHashMap();
+    rootJson.put("dist-tags", Maps.<String, Object>newHashMap());
+    rootJson.put("versions", Maps.<String, Object>newHashMap());
+    rootJson.put("x-nx-rebuilt", new Date());
+
+    for (Version version : packages.keySet()) {
+      try {
+        processPackageJson(rootJson, packages.get(version));
+      }
+      catch (IllegalArgumentException e) {
+        log.info("Failed to process package.json from {} version {}", coll.getPath(), version, e);
+      }
+    }
+
+    // ie. all processPackageJson failed?
+    if (((Map<String, Object>) rootJson.get("versions")).isEmpty()) {
+      return;
+    }
+
+    final Map<String, Object> lastPackage = packages.get(packages.lastKey());
+    final String lastVersion = String.valueOf(lastPackage.get("version"));
+    final Map<String, Object> distTags = (Map<String, Object>) rootJson.get("dist-tags");
+    distTags.remove(lastVersion);
+    distTags.put("latest", lastVersion);
+
+    final PackageRoot packageRoot = new PackageRoot(context.getRepository().getId(), rootJson);
+    hostedMetadataService.consumePackageRoot(packageRoot);
+    processedPackages = processedPackages + 1;
```

Comment:
is this really the number of processed packages, or the number of processed package roots? (or is there no difference in npm?)